R/binarise.R

```diff
@@ -0,0 +1,119 @@
+}
+
+binarise_GBXXX_to_GBXXXb_without_zero <- function(values) {
+    if("0" %in% values){
+        stop("Feature contains zero-values which are not permitted.")
+    }
+        dplyr::case_match(values, "1" ~ "0", "2" ~ "1", "3" ~ "1",  "?" ~ "?", NA ~ NA)
+}
+
+# functions for turning 2 of the multistate features into binarised version. These features have the 0 option.
+# we can just use this function for all multistate, since the other ones shouldn't legally have 0's in them at all. However, to be conservative I (Hedvig) separated them out so that if anything weird happens and somehow GB065 has a 0 value, the code breaks rather than does the wrong thing.
+
+#GB193 multistate 0: they cannot be used attributively, 1: ANM-N; 2: N-ANM; 3: both.
+#GB203 multistate 0: no UQ, 1: UQ-N; 2: N-UQ; 3: both.
+binarise_GBXXX_to_GBXXXa_with_zero <- function(values) {
+    dplyr::case_match(values, "0"~"0", "1" ~ "1", "2" ~ "0", "3" ~ "1", "?" ~ "?",  NA ~ NA)
+}
+
+binarise_GBXXX_to_GBXXXb_with_zero <- function(values) {
+    dplyr::case_match(values, "0"~"0", "1" ~ "0", "2" ~ "1", "3" ~ "1",  "?" ~ "?", NA ~ NA)
+}
+
+gb_recode <- function(ValueTable, oldvariable, newvariable, func) {
+    ValueTable %>% dplyr::filter(Parameter_ID==oldvariable) %>%
+        dplyr::mutate(
+            ID=paste0(newvariable, "-", Language_ID),
+            Parameter_ID=newvariable,
+            Value=func(Value)
+        ) %>%
+        dplyr::mutate(Code_ID = paste0(Parameter_ID, "-", Value)) %>%
+        rbind(ValueTable)
+}
+
+
+#' Makes multi-state Grambank-features binary in the appropriate manner.
+#'
+#' @param ValueTable data frame of the ValueTable from grambank-cldf (long).
+#' @param drop_multistate If TRUE,the multistate parent features of the binarised features are kept. If FALSE, they are dropped from the resulting data-frame.
+#' @param keep_raw_binary If TRUE and if the value table already contains some binarised features, they are kept. If false, they are overriden and replaced by values derived from the multistate features.
+#' @return Data-frame (long ValueTable)
+#' @export
+binarise <- function(ValueTable = NULL, drop_multistate = TRUE, keep_raw_binary = TRUE){
+    if (!inherits(ValueTable, "data.frame")) stop("'ValueTable' must be a dataframe.")
+
+    # if there are binary features coded directly by coders, i.e. not derived from multistate features,
+    # we call these "raw binary". users can choose to drop these and only used derived, or use a mix,
+    if (keep_raw_binary == FALSE) {
+        ValueTable <- ValueTable %>%
+            dplyr::filter(!(Parameter_ID %in% binary_parameters))
+
+    }
+
+    if (keep_raw_binary == TRUE) {
+        ValueTable_raw_binary <- ValueTable %>%
+            dplyr::filter(Parameter_ID %in% binary_parameters)
+
+    }
+
+
+
+# BINARISING MULTISTATE FEATUYES
+
+    ValueTable <- gb_recode(ValueTable, 'GB024', 'GB024a', binarise_GBXXX_to_GBXXXa_without_zero)
```

Comment:
Looking up the literal strings in `binary_parameters`/`multistate_parameters` would be more elegant and less prone to typos.

Reply:
sure. Ehm @SimonGreenhill do you want that?

Reply:
NB, I've split the set in two groups depending on wether they can take a 0 value or not. Perhaps a bit extra conservative, but I like sanity checks.

Reply:
Maybe a config dict like
```python
multistate_parameters = {
    'GB024': False,  # without zero
    ...
    'GB203': True,  # with zero
}
```
would do, then.

It just looked strange that a list of the literal feature IDs was defined above, and then not used down here.

Reply:
```python
for fid, with_zero in multistate_parameters.items():
    for ms in 'ab':
        ValueTable = gb_recode(ValueTable, fid, fid + ms, binarise_*_with_zero if with_zero else binarise_*_without_zero)
```

Reply:
@xrotwang I see what you mean. i use it in another place
@SimonGreenhill decides

Reply:
yes, agree that it's more elegant. I'll merge this now and make that a new issue (#16) for later refactoring.